### PR TITLE
stdc: add support for complex.h

### PR DESCRIPTION
**Pull request by lokesh-007**

### CODE REVIEW

**Comment by dspinellis** [comment]
The proposed change will not compile the following program.

```
#include <complex.h>

complex c;
```

**Comment by lokesh-007** [comment]
```
#define _Complex
#define complex _Complex 

```
and

```
#define _Complex
#define complex
```
 gives segmentation fault and is pretty obvious why! Defining a dummy macro is not working for the 'complex' keyword, however if we try

```
#define _Complex
#define complex long
 
```
works fine. 

But I fear that this is not a proper a way to do it, will it affect CScout's inference of the variable defined with 'complex'? Won't it be inferred as 'long'?

**Comment by dspinellis** [comment]
Exactly, you need to introduce a new type, for `_Complex`.

**Comment by lokesh-007** [comment]
I am sorry, But I am not able understand it, "introduce a new type for _Complex" what should be the type definition? are we going to use a custom defined struct or an inbuilt one?
Like 
```
#define _Complex <what-type?> 
or
typedef _Complex <what-type?> 

```
I checked in complex.h file in the library, there the line is
`#define complex _Complex`
which in itself, is in-built qualifier (https://www.gnu.org/software/libc/manual/html_node/Complex-Numbers.html) which I believe means that it comes from the language itself and not the library files. 

So, if we are going to simulate it, what should be our type definition?

**Comment by dspinellis** [comment]
Hi,

The modification needs to take place at the level of the CScout C++ code, not in a header file. This is also the case with long. 




<div>-------- Original message --------</div><div>From: Lokesh Prabakaran <notifications@github.com> </div><div>Date:04/04/2018  18:43  (GMT+02:00) </div><div>To: dspinellis/cscout <cscout@noreply.github.com> </div><div>Cc: Diomidis Spinellis <dds@aueb.gr>, Comment <comment@noreply.github.com> </div><div>Subject: Re: [dspinellis/cscout] stdc: add support for complex.h (#27) </div><div>
</div>I am sorry, But I am not able understand it, "introduce a new type for _Complex" what should be the type definition? are we going to use a custom defined struct or an inbuilt one?
Like

#define _Complex <what-type?> 
or
typedef _Complex <what-type?> 

I checked in complex.h file in the library, there the line is
#define complex _Complex
which in itself, is in-built qualifier (https://www.gnu.org/software/libc/manual/html_node/Complex-Numbers.html) which I believe means that it comes from the language itself and not the library files.

So, if we are going to simulate it, what should be our type definition?

**Comment by lokesh-007** [comment]
complex.h
```
#ifndef _COMPLEX_H
#define _COMPLEX_H	1
#define _Complex
typedef complex _Complex;
#endif /* complex.h */
```
this seems to work correctly, it complies 
#include <complex.h>
complex c;
float _Complex c1;

Does it look good? If its okie, I'll give a updated pull request. And why we should not do the modification at the header level? We are working only with the macros and typedef right?

**Comment by dspinellis** [comment]
This is a clever idea, but it still does not work for all cases. First according to the standard, `complex` has to be defined as a macro.
Second, the following will not compile.
```
_Complex a;
```
This is why I wrote that the change has to be made by modifying the CScout C++ source code.

**Comment by lokesh-007** [comment]
Consider this,
https://github.com/lokesh-007/cscout/commit/dae1e0025eee738757a5b26cf91018b846ae3396

Am I heading in the right direction? For a file like this,
```
#include <complex.h>
int main() {
  complex c; // will eventually become _Complex c;
}
```
the ctoken.cpp/parse_lex function correctly identifies '_Complex' as keyword with the value which I have given as 349.
I have observed that this is being called by parse.c/YYLex1 function, I am trying to wrap my head around it for some time,

What are those long array assignment like parse_lhs, parse_defred, etc? Any suggestions on how can I extend this parser to parse _Complex?

**Comment by dspinellis** [comment]
The `parse.c` file is generated from `parse.y` using *btyacc*.  The `parse.y`, rather than `parse.c` is the file you should be looking at.

**Comment by lokesh-007** [comment]
I have another doubt, how are eval.tab.h files and parse.tab.h files generated? 

The macro definition of COMPLEX goes there, which will be used by ctoken.cpp?

**Comment by lokesh-007** [comment]
I figured that parse.tab.h file's origin and have updated the code. It works fine and the tests have also been passed. Kindly check the updated changes.

**Comment by dspinellis** [comment]
Congratulations, you nailed it! I have only some very minor fix suggestions.

**Comment by lokesh-007** [comment]
I'll be happy to work on the changes! How do you think we can improve this?

File: include/stdc/complex.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+   <http://www.gnu.org/licenses/>.  */
+
+/*
+ *	ISO C99:  7.3 Complex arithmetic	<complex.h>
+ */
+
+#ifndef _COMPLEX_H
+#define _COMPLEX_H   1
+#define complex _Complex
+#endif /* complex.h */
```
[the remark this answers is not in the document]
Comment (dspinellis) [reply include/stdc/complex.h]: I added the (minor) comments as a code review. Please address, and update the pull request.

[the remark this answers is not in the document]
Comment (lokesh-007) [reply include/stdc/complex.h]: Can you share me a file, which has the license you are taking about? And what C99 standard specifications did I miss?

[the remark this answers is not in the document]
Comment (dspinellis) [reply include/stdc/complex.h]: Consider [this](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/complex.h.html) as a specification for `complex.h` and [this](https://github.com/dspinellis/cscout/blob/master/include/template/cscout-defs.h) as an example for a file with the project's license text.

**Comment by lokesh-007** [comment]
The changes have been made, kindly recheck the changes!

File: include/stdc/complex.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ *	ISO C99:  7.3 Complex arithmetic	<complex.h>
+ */
+
+#ifndef _COMPLEX_H
+#define _COMPLEX_H   1
+#define complex _Complex
+#define imaginary _Imaginary
+
+#define _Complex_I const float _Complex
+#define _Imaginary_I const float _Imaginary
```
Comment (dspinellis) [inline include/stdc/complex.h]: `_Imaginary_I` [should expand to a value](http://en.cppreference.com/w/c/numeric/complex/Imaginary_I) not a type.

File: include/stdc/complex.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+#define I _Imaginary_I
+
```
Comment (dspinellis) [inline include/stdc/complex.h]: You're missing the function declarations [listed in the standard](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/complex.h.html).

File: include/stdc/complex.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+/*
+ *	ISO C99:  7.3 Complex arithmetic	<complex.h>
+ */
+
+#ifndef _COMPLEX_H
+#define _COMPLEX_H   1
+#define complex _Complex
+#define imaginary _Imaginary
+
+#define _Complex_I const float _Complex
```
Comment (dspinellis) [inline include/stdc/complex.h]: `_Complex_I` [should expand to a value](http://en.cppreference.com/w/c/numeric/complex/Complex_I) not a type

File: include/stdc/complex.h [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+ *	ISO C99:  7.3 Complex arithmetic	<complex.h>
+ */
+
+#ifndef _COMPLEX_H
+#define _COMPLEX_H   1
+
+#define complex _Complex
+#define imaginary _Imaginary
+
+// setting it to -1f
```
Comment (dspinellis) [inline include/stdc/complex.h]: Remove unneeded comment.

File: include/stdc/complex.h [hunk 6]
```diff
@@ -0,0 +1,106 @@
+/*
+ * (C) Copyright 2002-2015 Diomidis Spinellis
```
Comment (dspinellis) [inline include/stdc/complex.h]: Put your name and 2018 here.

File: include/stdc/complex.h (hunk 5 above)
Comment (lokesh-007) [reply include/stdc/complex.h]: Done.

File: include/stdc/complex.h (hunk 6 above)
Comment (lokesh-007) [reply include/stdc/complex.h]: Done.

**Comment by lokesh-007** [comment]
I am not sure why the test fails, in my local when I ran 
`CFILES=c39-complex.c ./runtest.sh -p -TEST_C
`
I get a,
`./runtest.sh: line 457: ../../test/out/$i: ambiguous redirect
`
But this happens for any test file I ran. Do you think that this is causing the build to fail? I checked the logs, the reconstruction of chunk of complex file fails (i.e not ok) 

I have checked the err folder, all other files's .diff file seems empty except mine. 

Any suggestions?

**Comment by dspinellis** [comment]
If you look at the [Travis log](https://travis-ci.org/dspinellis/cscout/jobs/364929897) you will see that indeed the `complex.h` test case fails.  The `ambiguous redirect` message should not appear.  See what happens when I prime and run tests here:
```
$ CFILES=c12-call_graph.c ./runtest.sh -p -TEST_C
TAP version 13
Priming test data
1..0
# 0/0 tests passed
$ CFILES=c12-call_graph.c ./runtest.sh -TEST_C
TAP version 13
ok 1 - C c12-call_graph.c
1..1
# 1/1 tests passed
```
Are you using a non-standard shell?  Does your path contain a single quote in one of its elements?

You can trace why the error appears by running `CFILES=c39-complex.c bash -x runtest.sh -p -TEST_C`

**Comment by dspinellis** [comment]
It looks like priming is not compatible with running a single test.  Therefore run `./runtest.sh -p` and ensure that you only commit the files relevant to the test case you added.

**Comment by lokesh-007** [comment]
I ran all tests, I get the same output as the build machine in Travis, 

```
TAP version 13
Priming test data
ok 1 - reconstitution chunk c00-unused.c
ok 2 - reconstitution chunk c01-idtest.c
ok 3 - reconstitution chunk c02-include_sites.c
ok 4 - reconstitution chunk c03-ptrtypes.c
ok 5 - reconstitution chunk c04-designators.c
ok 6 - reconstitution chunk c05-gcc_labels.c
ok 7 - reconstitution chunk c06-namespace.c
ok 8 - reconstitution chunk c07-incomplete_typedef.c
ok 9 - reconstitution chunk c08-partial_identifiers.c
ok 10 - reconstitution chunk c09-concfun.c
ok 11 - reconstitution chunk c10-hello.c
ok 12 - reconstitution chunk c11-decldef.c
ok 13 - reconstitution chunk c12-call_graph.c
ok 14 - reconstitution chunk c13-allinc.c
ok 15 - reconstitution chunk c14-namespace.c
ok 16 - reconstitution chunk c15-fungraph.c
ok 17 - reconstitution chunk c16-cond.c
ok 18 - reconstitution chunk c17-inline.c
ok 19 - reconstitution chunk c18-typeof.c
ok 20 - reconstitution chunk c19-initializers.c
ok 21 - reconstitution chunk c20-implicit-fdecl.c
ok 22 - reconstitution chunk c21-chunking.c
ok 23 - reconstitution chunk c22-typeof.c
ok 24 - reconstitution chunk c23-opt-designator.c
ok 25 - reconstitution chunk c24-goto-expr.c
ok 26 - reconstitution chunk c25-nested-fun.c
ok 27 - reconstitution chunk c26-file-metric.c
ok 28 - reconstitution chunk c27-fun-metric.c
ok 29 - reconstitution chunk c28-fun-nparam.c
ok 30 - reconstitution chunk c29-spec-decl.c
ok 31 - reconstitution chunk c30-compile-const.c
ok 32 - reconstitution chunk c31-serial-initializers.c
ok 33 - reconstitution chunk c32-array-init.c
ok 34 - reconstitution chunk c33-incomplete-initializer.c
ok 35 - reconstitution chunk c34-idflink.c
ok 36 - reconstitution chunk c35-enum-designator.c
ok 37 - reconstitution chunk c36-endlabel.c
ok 38 - reconstitution chunk c37-yyvalid.c
ok 39 - reconstitution chunk c38-thread-local.c
not ok 40 - reconstitution chunk c39-complex.c
```
any idea why the priming is not compatible? If all works on your side, can you kindly add the test files from your side? If it is not too much trouble!

**Comment by dspinellis** [comment]
Reconstitution tests do not require priming, so it seems that you've successfully primed the data. (Please remember to add/commit the files.) Reconstitution tests parse the C file and then recreate it (reconstitute it) based on the database image of the file. Look at the difference between the files `test/c/c39-complex.c` and  `test/chunk/c39-complex.c`.  It should also appear in `test/err/chunk/c39-complex.c.diff` and the output of the test; not sure why it doesn't. In general, walk through the test script to see why tests fail.

**Comment by lokesh-007** [comment]
./test/c file
```
/* test program for complex.h support */

#include <complex.h>

int main() {
	complex c1;
	_Complex c2;
	float _Complex c3 = 2.0f + 2.0f*_Complex_I;

	imaginary c4;
	_Imaginary c5;
	float _Imaginary c7;

	float _Imaginary c8 = _Imaginary_I;
	float _Imaginary c9 = _Complex_I;
	float _Imaginary c10 = I;

	/*
	   Just an example; like this any function or macro function
	   declared in the standard complex.h when used, will be compiled by CScout
	*/
	double ca = cabs(c3);
	return 0;
}

```
./test/chunk/ file

`/* test program for complex.h support */\u0000a\u0000a#include <complex.h>\u0000a\u0000aint main() {\u0000a	complex c1;\u0000a	_Complex c2;\u0000a	float _Complex c3 = 2.0f + 2.0f*_Complex_I;\u0000a\u0000a	imaginary c4;\u0000a	_Imaginary c5;\u0000a	float _Imaginary c7;\u0000a\u0000a	float _Imaginary c8 = _Imaginary_I;\u0000a	float _Imaginary c9 = _Complex_I;\u0000a	float _Imaginary c10 = I;\u0000a\u0000a	/*\u0000a	   Just an example; like this any function or macro function\u0000a	   declared in the standard complex.h when used, will be compiled by CScout\u0000a	*/\u0000a	double ca = cabs(c3);\u0000a	return 0;\u0000a}\u0000a
`

other test files are generating normally, except this file? Any reasons?

**Comment by dspinellis** [comment]
I found and fixed the problem in the test script. Please retry after pulling in 57c34629c18de783a91077e4d5a7b72acbca35d9.

**Comment by lokesh-007** [comment]
Thanks for the update. The build went about good, kindly recheck the changes.

**Comment by dspinellis** [comment]
Congratulations for the merged pull request and kudos for your persistence!

**Pull request merged**